File: PRACTICAS-PDOO/IRRGARTEN/Versiones_en_Java_Ruby/PROYECTO3-JAVA/src/irrgarten/Game.java
/**
 * @file Game.java
 * @author angel_rodriguez
 */

package irrgarten;

import irrgarten.enums.GameCharacter;
import irrgarten.enums.Directions;
import irrgarten.enums.Orientation;
import java.util.ArrayList;
import java.util.Random;

public class Game {
    //Atributo de clase privado.
    private static final int MAX_ROUNDS = 10;
    
    //Atributos de instancias privados.
    private int currentPlayerIndex;     // índice del jugador actual.
    private String log;                 // cadena para almacenar el estado del juego.
    
    //Atributos de relación privados.
    private ArrayList <Player> players;      // contenedor de jugadores
    private Player currentPlayer;       // jugador actual
    private ArrayList <Monster> monsters;    // contenedor de monstruos
    private Labyrinth labyrinth;        // laberinto
    
    //Métodos de instancias públicos.
    
    /**
     * Constructor paramétrico.
     * @param nplayers número de jugadores.
     */
    public Game(int nplayers){
        /* SE DECIDE TODO DE FORMA ALEATORIA.
            -   Jugador que empieza la partida ---> currentPlayerIndex
            -   Número de monstruos.
            -   Posición de la casilla de salida (exitRow, exitCol)
        */
        
        Random generator = new Random();
        
        this.log = "";
        this.players = new ArrayList<Player>();
        this.monsters = new ArrayList<Monster>();
        
        // Creo y agrego los jugadores al contenedor.
        for (int i = 0; i < nplayers; i++){
            char numberPlayer = (char) ('0' + i);
            Player player = new Player(numberPlayer, 
                    Dice.randomIntelligence(), Dice.randomStrength());
            this.players.add(player);
        }
        
        this.currentPlayerIndex = generator.nextInt(nplayers);      
        this.currentPlayer = players.get(currentPlayerIndex);
        
        // El número de monstruos y la casilla de salida
        final int ROWS = 8; 
        final int COLS = 8;
        int 
            numMonster = generator.nextInt(10) + 1, //mínimo 1 monstruo.
            exitRow    = Dice.randomPos(ROWS),
            exitCol    = Dice.randomPos(COLS);
        
        // Añado los monstruos
        for( int i = 0; i < numMonster; i++){
            String nameMonster = "Monstruo " + i;
            Monster monster = new Monster(nameMonster, 
                    Dice.randomIntelligence(), Dice.randomStrength());
            this.monsters.add(monster);
        }
        
        // Instancio el laberinto
        this.labyrinth = new Labyrinth(ROWS, COLS, exitRow, exitCol);
        
        // Configuro el laberinto.
        configureLabyrinth();
        
        // Reparte los jugadores por el tablero.
        labyrinth.spreadPlayers(this.players);
    }
    
    /**
     * Constructor por defecto.
     */
    public Game(){ this(2);}
    
    public Game(int currentPlayerIndex, int nplayers, ArrayList<Player> players){
        this.currentPlayerIndex = 0;
        this.log = "";
        this.players = new ArrayList<Player>();
        
        // Creo y agrego los jugadores al contenedor.
        for (int i = 0; i < nplayers; i++){
            Player player = new Player();
            this.players.add(player);
        }
        
        this.currentPlayer = new Player();
        this.monsters = new ArrayList<Monster>();
        this.labyrinth = new Labyrinth();
    }
    
    public boolean finished(){
        return labyrinth.haveAWinner();        
    }
    
    public boolean nextStep(Directions preferredDirection){
        log = "";
        boolean dead = currentPlayer.dead();
        
        if(!dead){
            Directions direction = actualDirection(preferredDirection);
            
            if(direction != preferredDirection)
                logPlayerNoOrders();
            
            Monster monster = labyrinth.putPlayer(direction, currentPlayer);
            
            if(monster == null)
                logNoMonster();
            else{
                GameCharacter winner = combat(monster);
                manageReward(winner);
            }
            
        }else
            manageResurrection();
        
        boolean endGame = finished();
        
        if(!endGame)
            nextPlayer();
        
        return endGame;
    }
    
    /**
     * Devuelve un estado actual del juego.
     * @return GameState
     */
    public GameState getGameState(){
        String statePlayers = "", stateMonsters = "\n";
        for(int i = 0; i < players.size(); i++)
            statePlayers += players.get(i).toString() + "\n";
        for(int j = 0; j < monsters.size(); j++)
            stateMonsters += monsters.get(j).toString() + "\n";
        GameState gameState = new GameState(labyrinth.toString(), 
                statePlayers, stateMonsters,
                currentPlayerIndex, finished(), log);
        return gameState;
    }
    
    /**
     * Configura el laberinto añadiendo bloques e indicando la posición a
     * los monstruos.
     */
    private void configureLabyrinth(){
        Random generator = new Random();
        final int ROWS = 8; 
        final int COLS = 8;

        labyrinth.addBlock(Orientation.HORIZONTAL, 0, 0, 8);
        
        for(int i = 0; i < monsters.size(); i++){
          int row = generator.nextInt(ROWS),
              col = generator.nextInt(COLS);
            labyrinth.addMonster(row, col, monsters.get(i));
        }
    }

    
    private void nextPlayer(){
        this.currentPlayerIndex++;
        this.currentPlayer = this.players.get(currentPlayerIndex);
    }
    
    private Directions actualDirection(Directions preferredDirection){
        int currentRow = currentPlayer.getRow();
        int currentCol = currentPlayer.getCol();
        ArrayList<Directions> validMoves = labyrinth.validMoves(currentRow, 
                currentCol);
        Directions output = currentPlayer.move(preferredDirection, validMoves);
        
        return output;        
    }
    
    private GameCharacter combat(Monster monster){
        int rounds = 0;
        GameCharacter winner = GameCharacter.PLAYER;
        float playerAttack = currentPlayer.attack();
        boolean lose = monster.defend(playerAttack);
        while( (!lose) && (rounds < MAX_ROUNDS) ){
            winner = GameCharacter.MONSTER;
            rounds++;
            float monsterAttack = monster.attack();
            lose = currentPlayer.defend(monsterAttack);
            if(!lose){
                playerAttack = currentPlayer.attack();
                winner = GameCharacter.PLAYER;
                lose = monster.defend(playerAttack);
            }
        }
        
        logRounds(rounds, MAX_ROUNDS);
        
        return winner;    
    }
    
    private void manageReward(GameCharacter winner){
        if(winner == GameCharacter.PLAYER){
            currentPlayer.receiveReward();
            logPlayerWon();
        }else
            logMonsterWon();
    }
    
    private void manageResurrection(){
        boolean resurrect = Dice.resurrectPlayer();
        if(resurrect){
            currentPlayer.resurrect();
            logResurrected();
        }else
            logPlayerSkipTurn();
        
    }
    
    private void logPlayerWon(){
        this.log += "El jugador ha ganado el combate. \n";
    }
    
    private void logMonsterWon(){
        this.log += "El monstruo ha ganado el combate.\n";
    }
    
    private void logResurrected(){
        this.log += "El jugador ha resucitado.\n";
    }
    
    private void logPlayerSkipTurn(){
        this.log += "El jugador ha perdido el turno por estar muerto.\n";
    }
    
    private void logPlayerNoOrders(){
        this.log += "El jugador no ha podido seguir las instrucciones del jugador humano (no ha sido posible). \n";
    }
    
    private void logNoMonster(){
        this.log += "El jugador se ha movido a una celda vacía o no le ha sido posible moverse.\n";  
    }
    
    private void logRounds(int rounds, int max){
        this.log += "Se han producido " + rounds + " de " + max + " rondas de combate.\n";
    }
    
}

File: PRACTICAS-PDOO/IRRGARTEN/Versiones_en_Java_Ruby/PROYECTO3-RUBY/Irrgarten/game.rb
# frozen_string_literal: true
#encoding:utf-8

# File: game.rb
# Author: Ángel Rodríguez Faya

module Irrgarten
  class Game
    @@MAX_ROUNDS = 10.freeze


    public # Métodos públicos


    def initialize (nplayers)
      #SE DECIDE DE FORMA ALEATORIA.
      # -   Jugador que empieza la partida ---> currentPlayerIndex
      # -   Número de monstruos.
      # -   Posición de la casilla de salida (exitRow, exitCol)

      generator = Random.new

      @log = ""
      @players = Array.new
      @monsters = Array.new

      # Creo y agrego los jugadores al contenedor.
      for i in 0..(nplayers-1)
        number_player = i.chr
        player = Player.new(number_player, Dice.random_intelligence,
                            Dice.random_strength)
        @players.add(player)
      end

      @current_player_index = generator.rand(nplayers.to_i)
      @current_player = @players[@current_player_index]

      # El número de monstruos y la casilla de salida lo haré de forma aleatoria.
      rows = 8.freeze # Número de FILAS de mi laberinto.
      cols = 8.freeze # Número de COLUMNAS de mi laberinto.
      num_monster = generator.rand(10.to_i) + 1
      exit_row    = Dice.random_pos(rows)
      exit_col    = Dice.random_pos(cols)

      # Añado los monstruos
      for i in 0..(num_monster-1)
        name_monster = "Monstruo '" + i.chr + "'"
        monster = Monster.new(name_monster.to_i, Dice.random_intelligence,
                              Dice.random_strength)
        @monsters.add(monster)
      end

      # Instancio el laberinto.
      @labyrinth = Labyrinth.new(rows,cols,exit_row,exit_col)

    end

    def finished #boolean
      @labyrinth.have_a_winner
    end

    def next_step(preferredDirection) #boolean
      @log = ""
      dead = @current_player.dead

      if !dead
        direction = actual_direction(preferredDirection)

        if direction != preferredDirection
          log_player_no_orders
        end

        monster = @labyrinth.put_player(direction, @current_player)

        if monster == nil
          log_no_monster
        else
          winner = combat(monster)
          manage_reward(winner)
        end

      else
        manage_resurrection
      end

      end_game = finished

      if !end_game
        next_player
      end

      end_game
    end

    def get_game_state #GameState
      @game_state = GameState.new("", "", "", @current_player_index, finished, "")
      @game_state
    end


    private # Métodos privados


    def configure_labyrinth
    end

    def next_player
      @current_player_index += 1
      @current_player = @players.get(@current_player_index)
    end

    def actual_direction(preferred_direction) #Directions
      current_row = @current_player.row
      current_col = @current_player.col
      valid_moves = @labyrinth.valid_moves(current_row, current_col)
      output = @current_player.move(preferred_direction, valid_moves)
      output
    end

    def combat(monster) #GameCharacter
      rounds = 0
      winner = GameCharacter.PLAYER
      player_attack = @current_player.attack
      lose = monster.defend(player_attack)
      while !lose && rounds < @@MAX_ROUNDS
        winner = GameCharacter.MONSTER
        rounds += 1
        monster_attack = monster.attack
        lose = @current_player.defend(monster_attack)
        if !lose
          player_attack = @current_player.attack
          winner = GameCharacter.PLAYER
          lose = monster.defend(player_attack)
        end
      end
      log_rounds(rounds, @@MAX_ROUNDS)

      winner
    end

    def manage_reward(winner)
      if winner == GameCharacter.PLAYER
        @current_player.receive_reward
        log_player_won
      else
        log_monster_won
      end
    end

    def manage_resurrection
      resurrect = Dice.resurrect_player
      if resurrect
        @current_player.resurrect
        log_resurrected
      else
        log_player_skip_turn
      end
    end

    def log_player_won
      @log += "El jugador ha ganado el combate. \n"
    end

    def log_monster_won
      @log += "El monstruo ha ganado el combate.\n"
    end

    def log_resurrected
      @log += "El jugador ha resucitado.\n"
    end

    def log_player_skip_turn
      @log += "El jugador ha perdido el turno por estar muerto.\n"
    end

    def log_player_no_orders
      @log += "El jugador no ha podido seguir las instrucciones del jugador humano (no ha sido posible). \n"
    end

    def log_no_monster
      @log += "El jugador se ha movido a una celda vacía o no le ha sido posible moverse.\n"
    end

    def log_rounds(rounds, max)
      @log += "Se han producido #{rounds} de #{max} rondas de combate.\n"
    end

  end   # end de la clase Game
end  # end del módulo Irrgarten
